src/backend/distributed/commands/extension.c

```diff
@@ -115,3 +212,303 @@ ExtractNewExtensionVersion(Node *parsetree)
+		return false;
+	}
+
+	/* Generate the certificate. */
+	x509 = generate_x509(pkey);
+	if (!x509)
+	{
+		EVP_PKEY_free(pkey);
+		ereport(ERROR, (errmsg("error while generating x509 certificate")));
+		return false;
+	}
+
+	/* Write the private key and certificate out to disk. */
+	filesWritten = write_to_disk(pkey, x509);
+	EVP_PKEY_free(pkey);
+	X509_free(x509);
+
+	if (!filesWritten)
+	{
+		ereport(ERROR, (errmsg("error while writing key and certificate to file")));
+		return false;
+	}
+
+	return true;
+}
+
+
+/* Generates a 2048-bit RSA key. */
+static EVP_PKEY *
+generate_key_rsa()
+{
+	int ret = 0;
+	EVP_PKEY *pkey = NULL;
+	BIGNUM *bne = NULL;
+	RSA *rsa = NULL;
+
+	/* Allocate memory for the EVP_PKEY structure. */
+	pkey = EVP_PKEY_new();
+	if (!pkey)
+	{
+		ereport(ERROR, (errmsg("unable to create EVP_PLEY")));
+		return NULL;
+	}
+
+	/* 1. generate rsa key */
+	bne = BN_new();
+	ret = BN_set_word(bne, RSA_F4);
+	if (ret != 1)
+	{
+		EVP_PKEY_free(pkey);
+		ereport(ERROR, (errmsg("unable to prepare bignum")));
+		return NULL;
+	}
+
+	/* Generate the RSA key and assign it to pkey. */
+	rsa = RSA_new();
+	ret = RSA_generate_key_ex(rsa, 2048, bne, NULL);
+	BN_free(bne);
+	if (ret != 1)
+	{
+		EVP_PKEY_free(pkey);
+		ereport(ERROR, (errmsg("unable to generate RSA key")));
+		return NULL;
+	}
+
+	if (!EVP_PKEY_assign_RSA(pkey, rsa))
+	{
+		EVP_PKEY_free(pkey);
+		ereport(ERROR, (errmsg("unable to assign RSA key")));
+		return NULL;
+	}
+
+	/* The key has been generated, return it. */
+	return pkey;
+}
+
+
+static EVP_PKEY *
+generate_key_dsa()
+{
+	int ret = 0;
+	EVP_PKEY *pkey = NULL;
+	DSA *dsa = NULL;
+
+	/* Allocate memory for the EVP_PKEY structure. */
+	pkey = EVP_PKEY_new();
+	if (!pkey)
+	{
+		ereport(ERROR, (errmsg("unable to create EVP_PLEY")));
+		return NULL;
+	}
+
+	dsa = DSA_new();
+	ret = DSA_generate_parameters_ex(dsa, 1024, NULL, 0, NULL, NULL, NULL);
+	if (ret != 1)
+	{
+		EVP_PKEY_free(pkey);
+		ereport(ERROR, (errmsg("unable to generate DSA key parameters")));
+	}
+
+	ret = DSA_generate_key(dsa);
+	if (ret != 1)
+	{
+		EVP_PKEY_free(pkey);
+		ereport(ERROR, (errmsg("unable to generate DSA key")));
+	}
+
+	if (!EVP_PKEY_assign_DSA(pkey, dsa))
+	{
+		EVP_PKEY_free(pkey);
+		ereport(ERROR, (errmsg("unable to assign DSA key")));
+		return NULL;
+	}
+
+	/* The key has been generated, return it. */
+	return pkey;
+}
+
+
+/* Generates a self-signed x509 certificate. */
+static X509 *
+generate_x509(EVP_PKEY *pkey)
+{
+	X509 *x509 = NULL;
+	X509_NAME *name = NULL;
+
+	/* Allocate memory for the X509 structure. */
+	x509 = X509_new();
+	if (!x509)
+	{
+		ereport(ERROR, (errmsg("unable to create x509 cert")));
+		return NULL;
+	}
+
+	/* Set the serial number. */
+	ASN1_INTEGER_set(X509_get_serialNumber(x509), 1);
+
+	/* This certificate is valid from now until exactly one year from now. */
+	X509_gmtime_adj(X509_get_notBefore(x509), 0);
+	X509_gmtime_adj(X509_get_notAfter(x509), 31536000L);
```

Comment:
seems `X509_get_notAfter` and `X509_get_notBefore` are deprecated

https://www.openssl.org/docs/man1.1.0/crypto/X509_CRL_set1_lastUpdate.html

Does connectivity break after a year?

Reply:
I have been looking into this, an they are indeed deprecated in 1.1.0, however both my machine and travis are still using 1.0.x for openssl so we can't change this code.

There were other issues when postgres was compiled with 1.1.x which I have solved.